<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Good)</span></div>

## 1 Interrupting the flow

In [1]:
for i in range(10):
    if i > 6: break
    print(i)

0
1
2
3
4
5
6


In [2]:
for i in range(10):
    if i == 6: continue        # Don't proceed further
                               # IN THE CURRENT LOOP
                               # if i == 6
    print(i)

0
1
2
3
4
5
7
8
9


In [3]:
for i in range(10):
    if i % 2 == 0: continue     # Don't proceed if the remainder is zero
                                # I.e. if the number is even
    print(i)

1
3
5
7
9


In [4]:
number=0

while True:
    print(number)
    number += 1
    if number > 4: break

0
1
2
3
4


`break`= exit loop </br>`continue`= skip following steps for this loop and continue loop

## 2 List comprehension!

### 2.1 Basic syntax

In [6]:
[number for number in range(10)] # start from 0, 10 elements

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [7]:
[number**2 for number in range(10)]

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]

### 2.2 List comprehension with conditions

In [8]:
[number for number in range(10) if number % 2 ==0]

[0, 2, 4, 6, 8]

## 3 Other useful stuff

### 3.1 for with unpacking

In [9]:
x, y, z=[1, 2, 3]
print(f'x = {x}, y = {y}, z = {z}')


x = 1, y = 2, z = 3


In [10]:
py_superhero_info=[['Natasha Romanoff', 'Black Widow'],
                     ['Tony Stark', 'Iron Man'],
                     ['Stephen Strange', 'Doctor Strange']]

for real_name, super_name in py_superhero_info:
    print(f"{real_name} is Marvel's {super_name}!")

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.2 for with a zip()

In [11]:
super_names = ["Black Widow", "Iron Man", "Doctor Strange"]
real_names = ["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

for real_name, super_name in zip(real_names,super_names):
    print(f"{real_name} is Marvel's {super_name}!")


Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


### 3.3 for with dictionaries

In [12]:
superhero_info={"Natasha Romanoff": "Black Widow",
                "Tony Stark": "Iron Man",
                "Stephen Strange": "Doctor Strange"}

for key, value in superhero_info.items(): # items() splits dictionary into key and value
    print(f"{key} is Marvel's {value}!")


Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


In [13]:
for key in superhero_info.keys():
    value=superhero_info[key]
    print(f"{key} is Marvel's {value}!")

Natasha Romanoff is Marvel's Black Widow!
Tony Stark is Marvel's Iron Man!
Stephen Strange is Marvel's Doctor Strange!


## Exercise 1 :  Make me an odd list

In [22]:
list = []
for i in range(10):
    if i % 2 == 0: continue
    list.append(i**2)
print(list)

[1, 9, 25, 49, 81]


## Exercise 2 :  Make me another odd list

In [16]:
[i**2 for i in range(10) if i % 2 == 1]


[1, 9, 25, 49, 81]

## Exercise 3 :  Time me!

In [23]:
%%timeit
list = []
for i in range(10):
    if i % 2 == 0: continue
    list.append(i**2)

6.93 µs ± 134 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [24]:
%timeit [i**2 for i in range(10) if i % 2 == 1]

7.1 µs ± 390 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


The first method is faster (6.93 µs vs 7.1 µs)